File: apps/minecraft/velocity/src/main/java/crabcraft/net/crabUtilities/velocity/NicknameListener.java
package crabcraft.net.crabUtilities.velocity;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.connection.DisconnectEvent;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;
import redis.clients.jedis.JedisPubSub;

import java.util.UUID;

public class NicknameListener {

    public static final String HASH_KEY = "crabutilities:nicknames";
    public static final String UPDATE_CHANNEL = "crabutilities:nicknames-updates";

    private final CrabUtilitiesVelocity plugin;
    private final VelocityConfig config;
    private JedisPool jedisPool;
    private Thread subscriberThread;
    private JedisPubSub pubSub;
    private volatile boolean redisFailureLogged;

    public NicknameListener(CrabUtilitiesVelocity plugin, VelocityConfig config) {
        this.plugin = plugin;
        this.config = config;
    }

    public void start() {
        JedisPoolConfig poolConfig = new JedisPoolConfig();
        poolConfig.setMaxTotal(2);
        if (config.getRedisPassword() != null && !config.getRedisPassword().isEmpty()) {
            jedisPool = new JedisPool(poolConfig, config.getRedisHost(),
                    config.getRedisPort(), 2000, config.getRedisPassword());
        } else {
            jedisPool = new JedisPool(poolConfig, config.getRedisHost(),
                    config.getRedisPort(), 2000);
        }

        pubSub = new JedisPubSub() {
            @Override
            public void onMessage(String channel, String message) {
                if (UPDATE_CHANNEL.equals(channel)) ingest(message);
            }
        };

        subscriberThread = new Thread(() -> {
            boolean warned = false;
            while (!Thread.currentThread().isInterrupted()) {
                JedisPool pool = jedisPool;
                if (pool == null || pool.isClosed()) break;
                try (Jedis jedis = pool.getResource()) {
                    if (warned) {
                        plugin.getLogger().info("Nickname Redis subscriber reconnected.");
                        warned = false;
                    }
                    jedis.subscribe(pubSub, UPDATE_CHANNEL);
                } catch (NoClassDefFoundError e) {
                    break;
                } catch (Exception e) {
                    if (Thread.currentThread().isInterrupted()) break;
                    if (!warned) {
                        plugin.getLogger().warn("Nickname Redis subscriber unavailable, reconnecting in 3s...", e);
                        warned = true;
                    } else {
                        plugin.getLogger().debug("Nickname Redis subscriber disconnected: {}", e.getMessage());
                    }
                    try {
                        Thread.sleep(3000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                }
            }
        }, "CrabUtilities-Nickname-Subscriber");
        subscriberThread.setDaemon(true);
        subscriberThread.start();
    }

    public void publishNickname(UUID uuid, String raw) {
        JedisPool pool = jedisPool;
        if (pool == null || pool.isClosed()) return;

        String nickname = raw == null ? "" : raw;
        JsonObject payload = new JsonObject();
        payload.addProperty("uuid", uuid.toString());
        payload.addProperty("raw", nickname);

        plugin.getServer().getScheduler().buildTask(plugin, () -> {
            try (Jedis jedis = pool.getResource()) {
                jedis.hset(HASH_KEY, uuid.toString(), nickname);
                jedis.publish(UPDATE_CHANNEL, payload.toString());
                if (redisFailureLogged) {
                    plugin.getLogger().info("Nickname Redis publisher recovered.");
                    redisFailureLogged = false;
                }
            } catch (Exception e) {
                if (!redisFailureLogged) {
                    plugin.getLogger().warn("Failed to publish nickname for {}; will retry on later updates", uuid, e);
                    redisFailureLogged = true;
                } else {
                    plugin.getLogger().debug("Failed to publish nickname for {}: {}", uuid, e.getMessage());
                }
            }
        }).schedule();
    }

    public String loadRawNickname(UUID uuid) {
        JedisPool pool = jedisPool;
        if (pool == null || pool.isClosed()) return null;

        try (Jedis jedis = pool.getResource()) {
            String raw = jedis.hget(HASH_KEY, uuid.toString());
            if (redisFailureLogged) {
                plugin.getLogger().info("Nickname Redis reader recovered.");
                redisFailureLogged = false;
            }
            return raw;
        } catch (Exception e) {
            if (!redisFailureLogged) {
                plugin.getLogger().warn("Failed to read nickname for {} from Redis", uuid, e);
                redisFailureLogged = true;
            } else {
                plugin.getLogger().debug("Failed to read nickname for {} from Redis: {}", uuid, e.getMessage());
            }
            return null;
        }
    }

    private void ingest(String json) {
        final UUID uuid;
        final String raw;
        try {
            JsonObject obj = JsonParser.parseString(json).getAsJsonObject();
            uuid = UUID.fromString(obj.get("uuid").getAsString());
            raw = obj.has("raw") && !obj.get("raw").isJsonNull()
                    ? obj.get("raw").getAsString()
                    : "";
        } catch (Exception e) {
            plugin.getLogger().warn("Ignoring malformed nickname Redis update", e);
            return;
        }

        if (plugin.getServer().getPlayer(uuid).filter(player -> player.isActive()).isEmpty()) {
            return;
        }

        plugin.getNicknameCache().setNickname(uuid, raw);
        plugin.getPendingJoinManager().complete(uuid);
        persist(uuid);
    }

    @Subscribe
    public void onDisconnect(DisconnectEvent event) {
        UUID uuid = event.getPlayer().getUniqueId();
        plugin.getPendingJoinManager().remove(uuid);
        plugin.getNicknameCache().remove(uuid);
        if (plugin.getMessageManager() != null) {
            plugin.getMessageManager().clearReplyTargets(uuid);
            plugin.getMessageManager().clearSpy(uuid);
        }
    }

    public void shutdown() {
        if (pubSub != null) {
            try {
                pubSub.unsubscribe();
            } catch (Exception ignored) {}
        }
        if (subscriberThread != null) {
            subscriberThread.interrupt();
            try {
                subscriberThread.join(2000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
        if (jedisPool != null && !jedisPool.isClosed()) {
            try {
                jedisPool.close();
            } catch (NoClassDefFoundError ignored) {}
        }
        jedisPool = null;
    }

    void persist(UUID uuid) {
        final String uuidStr = uuid.toString();
        final String plain = plugin.getNicknameCache().getPlainNickname(uuid);
        final String raw = plugin.getNicknameCache().getRawNickname(uuid);
        plugin.runDatabaseTask("nickname-persist",
                () -> plugin.getPgWriter().updateNickname(uuidStr, plain, raw));
    }
}

File: apps/minecraft/velocity/src/main/kotlin/crabcraft/net/crabUtilities/velocity/NicknameListener.kt
package crabcraft.net.crabUtilities.velocity

import com.google.gson.JsonParser
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.connection.DisconnectEvent
import crabcraft.net.crabUtilities.redis.RedisSubscriberThread
import java.util.UUID
import redis.clients.jedis.JedisPool
import redis.clients.jedis.JedisPubSub

open class NicknameListener(private val plugin: CrabUtilitiesVelocity, private val config: VelocityConfig) {
    private var jedisPool: JedisPool? = null
    private var subscriberThread: Thread? = null
    private var pubSub: JedisPubSub? = null
    @Volatile private var redisFailureLogged = false

    open fun start() {
        jedisPool = RedisPools.create(config, 2)
        pubSub =
            object : JedisPubSub() {
                override fun onMessage(channel: String, message: String) {
                    if (UPDATE_CHANNEL == channel) ingest(message)
                }
            }
        subscriberThread =
            RedisSubscriberThread.start(
                "CrabUtilities-Nickname-Subscriber",
                { jedisPool },
                { jedis -> jedis.subscribe(pubSub, UPDATE_CHANNEL) },
                { plugin.getLogger().info("Nickname Redis subscriber reconnected.") },
                { error, firstFailure ->
                    if (firstFailure) {
                        plugin.getLogger().warn("Nickname Redis subscriber unavailable, reconnecting in 3s...", error)
                    } else plugin.getLogger().debug("Nickname Redis subscriber disconnected: {}", error.message)
                },
            )
    }

    open fun publishNickname(uuid: UUID, raw: String?, expectedVersion: Long) {
        val pool = jedisPool
        if (pool == null || pool.isClosed) return
        val nickname = raw ?: ""
        plugin
            .getServer()
            .scheduler
            .buildTask(
                plugin,
                Runnable {
                    if (!plugin.getNicknameCache().isVersion(uuid, expectedVersion)) return@Runnable
                    try {
                        pool.resource.use { jedis ->
                            val actual =
                                jedis.eval(
                                    PUBLISH_IF_ABSENT_SCRIPT,
                                    listOf(HASH_KEY),
                                    listOf(uuid.toString(), nickname, UPDATE_CHANNEL),
                                ) as String?
                            if (
                                reconcilePublishedNickname(
                                    plugin.getNicknameCache(),
                                    uuid,
                                    expectedVersion,
                                    nickname,
                                    actual,
                                )
                            ) {
                                persist(uuid)
                            }
                            if (redisFailureLogged) {
                                plugin.getLogger().info("Nickname Redis publisher recovered.")
                                redisFailureLogged = false
                            }
                        }
                    } catch (error: Exception) {
                        if (!redisFailureLogged) {
                            plugin
                                .getLogger()
                                .warn("Failed to publish nickname for {}; will retry on later updates", uuid, error)
                            redisFailureLogged = true
                        } else plugin.getLogger().debug("Failed to publish nickname for {}: {}", uuid, error.message)
                    }
                },
            )
            .schedule()
    }

    open fun loadRawNickname(uuid: UUID): String? {
        val pool = jedisPool
        if (pool == null || pool.isClosed) return null
        return try {
            pool.resource.use { jedis ->
                val raw = jedis.hget(HASH_KEY, uuid.toString())
                if (redisFailureLogged) {
                    plugin.getLogger().info("Nickname Redis reader recovered.")
                    redisFailureLogged = false
                }
                raw
            }
        } catch (error: Exception) {
            if (!redisFailureLogged) {
                plugin.getLogger().warn("Failed to read nickname for {} from Redis", uuid, error)
                redisFailureLogged = true
            } else plugin.getLogger().debug("Failed to read nickname for {} from Redis: {}", uuid, error.message)
            null
        }
    }

    private fun ingest(json: String) {
        val uuid: UUID
        val raw: String
        try {
            val obj = JsonParser.parseString(json).asJsonObject
            uuid = UUID.fromString(obj.get("uuid").asString)
            raw = if (obj.has("raw") && !obj.get("raw").isJsonNull) obj.get("raw").asString else ""
        } catch (error: Exception) {
            plugin.getLogger().warn("Ignoring malformed nickname Redis update", error)
            return
        }
        if (plugin.getServer().getPlayer(uuid).filter { it.isActive }.isEmpty) return
        plugin.getNicknameCache().setNickname(uuid, raw)
        plugin.getPendingJoinManager().complete(uuid)
        persist(uuid)
    }

    @Subscribe
    open fun onDisconnect(event: DisconnectEvent) {
        val uuid = event.player.uniqueId
        plugin.getPendingJoinManager().remove(uuid)
        plugin.getNicknameCache().remove(uuid)
        plugin.getMessageManager()?.let {
            it.clearReplyTargets(uuid)
            it.clearSpy(uuid)
        }
    }

    open fun shutdown() {
        pubSub?.let {
            try {
                it.unsubscribe()
            } catch (ignored: Exception) {}
        }
        subscriberThread?.let {
            it.interrupt()
            try {
                it.join(2000)
            } catch (error: InterruptedException) {
                Thread.currentThread().interrupt()
            }
        }
        jedisPool?.let {
            if (!it.isClosed)
                try {
                    it.close()
                } catch (ignored: NoClassDefFoundError) {}
        }
        jedisPool = null
    }

    fun persist(uuid: UUID) {
        val uuidString = uuid.toString()
        val plain = plugin.getNicknameCache().getPlainNickname(uuid)
        val raw = plugin.getNicknameCache().getRawNickname(uuid)
        plugin.runDatabaseTask("nickname-persist") { plugin.getPgWriter()!!.updateNickname(uuidString, plain, raw) }
    }

    companion object {
        const val HASH_KEY = "crabutilities:nicknames"
        const val UPDATE_CHANNEL = "crabutilities:nicknames-updates"
        private val PUBLISH_IF_ABSENT_SCRIPT =
            """
            local current = redis.call('HGET', KEYS[1], ARGV[1])
            if current == false then
                current = ARGV[2]
                redis.call('HSET', KEYS[1], ARGV[1], current)
            end
            local payload = cjson.encode({uuid = ARGV[1], raw = current})
            redis.call('PUBLISH', ARGV[3], payload)
            return current
            """
                .trimIndent() + "\n"

        @JvmStatic
        fun reconcilePublishedNickname(
            cache: NicknameCache,
            uuid: UUID,
            expectedVersion: Long,
            proposed: String,
            actual: String?,
        ): Boolean = proposed != actual && cache.commitIfVersion(uuid, expectedVersion, actual)
    }
}
